_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,507 @@
+    }
+
+    @Bean
+    public ConcurrentKafkaListenerContainerFactory<String, User> userKafkaListenerContainerFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, User> factory = new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(userConsumerFactory());
+        return factory;
+    }
+    ...
+}
+```
+
+Spring Kafka provides `JsonSerializer` and `JsonDeserializer` implementations that are based on the Jackson JSON object mapper. It allows us to convert any Java Object to `bytes[]`.
+
+In the above example, we are creating one more `ConcurrentKafkaListenerContainerFactory` for JsonSerialization. In this, we have configured `JsonSerializer.class` as our value serializer in Producer config and `JsonDeserializer<>(User.class)` as out value deserializer in consumer config. For this, we are creating separate Kafka listener container `userKafkaListenerContainerFactory`. If we have multiple Java object types to be serialized/deserialized, we have to create a listener container for each type as shown above.
+
+Now that we have configured our serializer & deserializer, we can send User object using `KafkaTemplate`.
+
+```java
+@Component
+class KafkaSenderExample {
+  ...
+
+  @Autowired
+  private KafkaTemplate<String, User> userKafkaTemplate;
+
+  void sendCustomMessage(User user, String topicName) {
+    userKafkaTemplate.send(topicName, user);
+  }
+  ...
+}
+```
+
+As shown above, we can create `KafkaTemplate` for the specific type and send user records to designated topics.
+
+```java
+@Component
+class KafkaListenersExample {
+
+  @KafkaListener(topics = "reflectoring-user", groupId="reflectoring-user", containerFactory="userKafkaListenerContainerFactory")
+  void listener(User user) {
+    LOG.info("CustomUserListener [{}]", user);
+  }
+}
+```
+
+Since we have multiple listener containers, we are specifying which container factory (`containerFactory="userKafkaListenerContainerFactory"`) to be used while consuming records.
+
+If we don't specify this attribute it defaults to `kafkaListernerContainerFactory` which uses `StringSerializer` and `StringDeserializer` in our case.
+
+### Using Spring Messaging Message Conversion
+
+```java
+@Configuration
+class KafkaProducerConfig {
+  ...
+
+  @Bean
+  KafkaTemplate<String, String> kafkaTemplate() {
+    ...
+    kafkaTemplate.setMessageConverter(new StringJsonMessageConverter());
+    kafkaTemplate.setDefaultTopic("reflectoring-user");
+    return kafkaTemplate;
+  }
+  ...
+}
+```
+
+Spring Kafka provides `MessageConverter` abstractions with `JsonMessageConverter`. We can inject `MessageConverter` into our `KafkaTemplate` instance directly. We must configure a message converter that is compatible with the configured Kafka Serializer. We are using `StringJsonMessageConverter` with `StringSerializer`. Also, a default topic must to set, becaause we'll be separating records based on the type of data we are sending.
```

Comment:
Where is the `StringJsonSerializer`? I don't see it in the code.

Reply:
```java
@Bean
KafkaTemplate<String, String> kafkaTemplate() {
	KafkaTemplate<String, String> kafkaTemplate = new KafkaTemplate<>(producerFactory());
	kafkaTemplate.setMessageConverter(new StringJsonMessageConverter());
	kafkaTemplate.setDefaultTopic("reflectoring-user");
}
```
We are using `producerFactory()` and Serializer is defined there